### PR TITLE
MultiReduceDiags.cpp : use a map to replace a long list of "if else" to choose a reduced diag

### PR DESCRIPTION
**Pull request by lucafedeli88**
The list of available reduced diags is growing! This PR uses a map to replace a long list of "if else" statements.

### CODE REVIEW

**Review by Yin-YinjianZhao — APPROVED** [review]

File: Source/Diagnostics/ReducedDiags/MultiReducedDiags.cpp [hunk 1]
```diff
@@ -35,75 +35,34 @@ MultiReducedDiags::MultiReducedDiags ()
     // if names are not given, reduced diags will not be done
     if ( m_plot_rd == 0 ) { return; }
 
-    // resize
-    m_multi_rd.resize(m_rd_names.size());
-
+    using CS = const std::string& ;
+    const auto reduced_diags_dictionary =
+        std::map<std::string, std::function<std::unique_ptr<ReducedDiags>(CS)>>{
```
Comment (ax3l) [inline Source/Diagnostics/ReducedDiags/MultiReducedDiags.cpp:44]: Nice! Just needs a few new includes :-)
- `<map>`
- `<algorithm>`
- `<string>`
- `<functional>`
- `<memory>`
- `<iterator>`

Comment (lucafedeli88) [reply Source/Diagnostics/ReducedDiags/MultiReducedDiags.cpp:44]: Yeah, that's right ! `string` and `memory` are already in `MultiReducedDiags.H`, but the others are missing...

Comment (lucafedeli88) [reply Source/Diagnostics/ReducedDiags/MultiReducedDiags.cpp:44]: I should have just added them!

File: Source/Diagnostics/ReducedDiags/MultiReducedDiags.cpp [hunk 2]
```diff
@@ -35,75 +39,34 @@ MultiReducedDiags::MultiReducedDiags ()
...
+            {"FieldMaximum",          [](CS s){return std::make_unique<FieldMaximum>(s);}},
+            {"RhoMaximum",            [](CS s){return std::make_unique<RhoMaximum>(s);}},
+            {"BeamRelevant",          [](CS s){return std::make_unique<BeamRelevant>(s);}},
+            {"LoadBalanceCosts",      [](CS s){return std::make_unique<LoadBalanceCosts>(s);}},
+            {"LoadBalanceEfficiency", [](CS s){return std::make_unique<LoadBalanceEfficiency>(s);}},
+            {"ParticleHistogram",     [](CS s){return std::make_unique<ParticleHistogram>(s);}},
+            {"ParticleNumber",        [](CS s){return std::make_unique<ParticleNumber>(s);}},
+            {"ParticleExtrema",       [](CS s){return std::make_unique<ParticleExtrema>(s);}}
+        };
     // loop over all reduced diags
```
Comment (NeilZaim) [inline Source/Diagnostics/ReducedDiags/MultiReducedDiags.cpp]:
```suggestion
    // loop over all reduced diags and fill m_multi_rd with requested reduced diags
```
I guess making this slightly more explicit could be useful for those of us not too familiar with the STL. :)

**Review by NeilZaim — APPROVED** [review]
Thanks for this PR, this is better indeed!

I just had a very minor suggestion.

**Review by ax3l — APPROVED** [review]
Thank you! :)

**Pull request merged**